apps/prairielearn/python/python_helper_sympy.py

```diff
@@ -225,64 +225,91 @@ class HasInvalidSymbolError(BaseSympyError):
         self.generic_visit(node)
 
-
-class CheckVariables(ast.NodeVisitor):
-    def __init__(self, variables: SympyMapT, functions: SympyMapT) -> None:
-        # functions is only used for error type, if someone writes "exp + 2"
-        self.variables = variables
-        self.functions = functions
-
     def visit_Name(self, node: ast.Name) -> None:
         if (
             isinstance(node.ctx, ast.Load)
-            and not is_name_of_function(node)
+            and not self.is_name_of_function(node)
             and node.id not in self.variables
         ):
-            err_node = get_parent_with_location(node)
+            err_node = self.get_parent_with_location(node)
             if node.id in self.functions:
-                raise FunctionNameUsedWithoutArguments(err_node.col_offset, err_node.id)
+                raise FunctionNameWithoutArgumentsError(
+                    err_node.col_offset, err_node.id
+                )
             else:
                 raise HasInvalidVariableError(err_node.col_offset, err_node.id)
         self.generic_visit(node)
 
+    def is_name_of_function(self, node: ast.AST) -> bool:
+        # The node is the name of a function if all of the following are true:
+        # 1) it has type ast.Name
+        # 2) its parent has type ast.Call
+        # 3) it is not in the list of parent's args
+        if not isinstance(node, ast.Name):
+            return False
+
+        parent = self.__parents[id(node)]
 
-def is_name_of_function(node: ast.AST) -> bool:
-    # The node is the name of a function if all of the following are true:
-    # 1) it has type ast.Name
-    # 2) its parent has type ast.Call
-    # 3) it is not in the list of parent's args
-    return isinstance(node, ast.Name) and isinstance(node.parent, ast.Call) and (node not in node.parent.args)  # type: ignore
+        return isinstance(parent, ast.Call) and (node not in parent.args)
 
+    def get_parent_with_location(self, node: ast.AST) -> Any:
+        while id(node) in self.__parents:
+            if hasattr(node, "col_offset"):
+                return node
+
+            node = self.__parents[id(node)]
 
-def get_parent_with_location(node: ast.AST) -> Any:
-    if hasattr(node, "col_offset"):
         return node
 
-    return get_parent_with_location(node.parent)  # type: ignore
+    def check_expression(self, expr: str) -> None:
+        # Parse (convert string to AST)
+        try:
+            root = ast.parse(expr, mode="eval")
+        except SyntaxError as err:
+            offset = err.offset if err.offset is not None else -1
+            raise HasParseError(offset)
+
+        # Link each node to its parent
+        self.__parents = {
+            id(child): node
+            for node in ast.walk(root)
+            for child in ast.iter_child_nodes(node)
+        }
+
+        self.visit(root)
+
+        # Empty parents dict after execution
+        # dict is only populated during execution
+        self.__parents = dict()
```

Comment:
Is this necessary since we'll always re-initialize `self.__parents` before calling `self.visit(root)`?

Reply:
Not strictly necessary, but just to avoid storing unnecessary data after it's done being used and maintain consistency with what's set in the `__init__` function.